### ISSUE TITLE
[r] Support R Conda installs

### ISSUE THREAD

**Issue by johnkerl** [single-cell-data/TileDB-SOMA#1277]
cc @aaronwolen @Shelnutt2 @eddelbuettel 

* [X] Remove the use of the special build of C++, i.e. remove the use of the cmake "build for R" option and remove the use of the `rcppspdlog` header. R has to use the regular SOMA C++ library.
* [X] Link against the shared library not static: #1293.
* [X] Remove the use of the bundled `.tar.gz` of the source code. No need for this, i.e TileDB-R doesn't bundle a `.tar.gz` of TileDB core C++ code.
* [ ] Add to the feedstock after the above is done.

**Comment by Shelnutt2** [comment]
> 2. Sure when one exists it will be easy to adapt the link process. That is actually independent of Conda use.

Not quite independent because we must use the common shared library built by conda (libtiledbsoma). So the R package for SOMA needs to be able to support linking against the shared library we have in the env. This is is what this is referring to.

**Comment by eddelbuettel** [comment]
Someone would need to provide a shared library the R package can build against.  Currently the build instructions (eg [here](https://github.com/single-cell-data/TileDB-SOMA/blob/main/libtiledbsoma/README.md#developer-makefile) have no such shared library.

**Comment by eddelbuettel** [comment]
To witness, when I allow `bld` (which is still outside `api/r`) to populate files below /usr/local the build (after copying over headers, cmake files, pkgconfig files, ...) eventually fails on a missing link for `fmt::format()`:

```sh
Install the project...
-- Install configuration: "Release"
-- Up-to-date: /usr/local/include/span
-- Up-to-date: /usr/local/include/span/span.hpp
-- Up-to-date: /usr/local/include/thread_pool
-- Up-to-date: /usr/local/include/thread_pool/thread_pool.h
-- Up-to-date: /usr/local/include/thread_pool/producer_consumer_queue.h
-- Up-to-date: /usr/local/include/thread_pool/status.h
-- Up-to-date: /usr/local/include/tiledbsoma
-- Up-to-date: /usr/local/include/tiledbsoma/array_buffers.h
-- Up-to-date: /usr/local/include/tiledbsoma/tiledbsoma
-- Up-to-date: /usr/local/include/tiledbsoma/managed_query.h
-- Up-to-date: /usr/local/include/tiledbsoma/util.h
-- Up-to-date: /usr/local/include/tiledbsoma/logger_public.h
-- Up-to-date: /usr/local/include/tiledbsoma/arrow_adapter.h
-- Up-to-date: /usr/local/include/tiledbsoma/column_buffer.h
-- Up-to-date: /usr/local/include/tiledbsoma/stats.h
-- Up-to-date: /usr/local/include/tiledbsoma/common.h
-- Up-to-date: /usr/local/include/tiledbsoma/soma_array_reader.h
-- Up-to-date: /usr/local/include/tiledbsoma/spdlog_with_R.h
-- Up-to-date: /usr/local/include/tiledbsoma/version.h
-- Up-to-date: /usr/local/include/tiledbsoma/carrow.h
-- Installing: /usr/local/lib/cmake/tiledbsoma/TileDBSomaConfig.cmake
-- Installing: /usr/local/lib/libtiledbsoma.so
-- Set runtime path of "/usr/local/lib/libtiledbsoma.so" to "/usr/local/lib"
-- Installing: /usr/local/include/tiledbsoma/tiledbsoma_export.h
-- Installing: /usr/local/bin/tdbsoma
-- Set runtime path of "/usr/local/bin/tdbsoma" to "/usr/local/lib"
-- Installing: /usr/local/lib/cmake/tiledbsoma/TileDBSomaTargets.cmake
-- Installing: /usr/local/lib/cmake/tiledbsoma/TileDBSomaTargets-release.cmake
-- Installing: /usr/local/lib/pkgconfig/tiledbsoma.pc
Built target install-libtiledbsoma
[  7%] Built target TILEDB_SOMA_OBJECTS
[ 93%] Built target Catch2
[ 95%] Built target Catch2WithMain
[ 97%] Building CXX object test/CMakeFiles/unit_soma.dir/unit_managed_query.cc.o
[ 97%] Building CXX object test/CMakeFiles/unit_soma.dir/unit_soma_array_reader.cc.o
[ 98%] Building CXX object test/CMakeFiles/unit_soma.dir/unit_column_buffer.cc.o
[100%] Linking CXX executable unit_soma
/usr/bin/ld: CMakeFiles/unit_soma.dir/unit_column_buffer.cc.o: undefined reference to symbol '_ZN3fmt2v96detail9dragonbox10to_decimalIfEENS2_10decimal_fpIT_EES5_'
/usr/bin/ld: /lib/x86_64-linux-gnu/libfmt.so.9: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
gmake[3]: *** [test/CMakeFiles/unit_soma.dir/build.make:146: test/unit_soma] Error 1
gmake[2]: *** [CMakeFiles/Makefile2:236: test/CMakeFiles/unit_soma.dir/all] Error 2
gmake[1]: *** [CMakeFiles/Makefile2:269: test/CMakeFiles/build_tests.dir/rule] Error 2
gmake: *** [Makefile:231: build_tests] Error 2
edd@rob:~/git/tiledb-soma/apis/r/src/libtiledbsoma(de/no_r_build_flag)$ 
edd@rob:~/git/tiledb-soma/apis/r/src/libtiledbsoma(de/no_r_build_flag)$ c++filt _ZN3fmt2v96detail9dragonbox10to_decimalIfEENS2_10decimal_fpIT_EES5_
fmt::v9::detail::dragonbox::decimal_fp<float> fmt::v9::detail::dragonbox::to_decimal<float>(float)
edd@rob:~/git/tiledb-soma/apis/r/src/libtiledbsoma(de/no_r_build_flag)$ 
```

This does not seem to be something caused by, or related to, the tiledbsoma R package.

**Comment by eddelbuettel** [comment]
For completeness that attempt also included these minimal changes to the build files to no do anything R specific:

```diff
modified   apis/r/src/Makevars.in
@@ -1,7 +1,8 @@
 CXX_STD = CXX17
 
 ## We need the TileDB Headers, and for macOS aka Darwin need to set minimum version 10.14 for macOS
-PKG_CPPFLAGS = -I. -I../inst/include/ @tiledb_include@ -DR_BUILD @cxx17_macos@ -Iinclude
+#PKG_CPPFLAGS = -I. -I../inst/include/ @tiledb_include@ -DR_BUILD @cxx17_macos@ -Iinclude
+PKG_CPPFLAGS = -I. -I../inst/include/ @tiledb_include@ @cxx17_macos@ -Iinclude
 
 ## We also need the TileDB library
 PKG_LIBS = @cxx17_macos@ @tiledb_libs@ @tiledb_rpath@ lib/libtiledbsoma.a
modified   libtiledbsoma/cmake/Modules/FindTileDB_EP.cmake
@@ -132,7 +132,8 @@ else()
   endif()
 endif()
 
-if (EP_TILEDB_BUILT AND TARGET TileDB::tiledb_shared AND NOT TILEDBSOMA_BUILD_R)
+#if (EP_TILEDB_BUILT AND TARGET TileDB::tiledb_shared AND NOT TILEDBSOMA_BUILD_R)
+if (EP_TILEDB_BUILT AND TARGET TileDB::tiledb_shared)
   include(TileDBCommon)
   install_target_libs(TileDB::tiledb_shared)
 endif()
modified   libtiledbsoma/cmake/Superbuild.cmake
@@ -67,7 +67,7 @@ set(INHERITED_CMAKE_ARGS
   -DSANITIZER=${SANITIZER}
   -DENABLE_ARROW_EXPORT=${ENABLE_ARROW_EXPORT}
   -DOVERRIDE_INSTALL_PREFIX=${OVERRIDE_INSTALL_PREFIX}
-  -DTILEDBSOMA_BUILD_R=${TILEDBSOMA_BUILD_R}
+  #-DTILEDBSOMA_BUILD_R=${TILEDBSOMA_BUILD_R}
   -DTILEDBSOMA_BUILD_STATIC=${TILEDBSOMA_BUILD_STATIC}
   -DTILEDBSOMA_BUILD_CLI=${TILEDBSOMA_BUILD_CLI}
   -DTILEDBSOMA_ENABLE_TESTING=${TILEDBSOMA_ENABLE_TESTING}

```

**Comment by eddelbuettel** [comment]
I can coax it work (at least locally) by telling `spdlog` to _not_ link with `libfmt`, and with the as-usual-missing `sudo ldconfig` step:

```diff
modified   apis/r/src/Makevars.in
@@ -5,7 +5,8 @@ CXX_STD = CXX17
 PKG_CPPFLAGS = -I. -I../inst/include/ @tiledb_include@ @cxx17_macos@ -Iinclude
 
 ## We also need the TileDB library
-PKG_LIBS = @cxx17_macos@ @tiledb_libs@ @tiledb_rpath@ lib/libtiledbsoma.a
+#PKG_LIBS = @cxx17_macos@ @tiledb_libs@ @tiledb_rpath@ lib/libtiledbsoma.a
+PKG_LIBS = @cxx17_macos@ -ltiledbsoma @tiledb_libs@ @tiledb_rpath@
 
 all: $(SHLIB)
         # if we are
modified   libtiledbsoma/cmake/Modules/FindSpdlog_EP.cmake
@@ -121,7 +121,8 @@ if (spdlog_FOUND AND NOT TARGET spdlog::spdlog)
 elseif(TARGET spdlog::spdlog)
   if (SPDLOG_FMT_EXTERNAL)
     # Since we are using header only we need to define this
-    add_definitions("-DSPDLOG_FMT_EXTERNAL=1")
+    # cf https://github.com/gabime/spdlog/wiki/0.-FAQ#how-to-use-external-fmt-library-instead-of-the-bundled
+    add_definitions("-DSPDLOG_FMT_EXTERNAL_HO=1")
     find_package(fmt REQUIRED)
     if (${fmt_FOUND})
       target_link_libraries(spdlog::spdlog INTERFACE fmt::fmt)

```

Then:

```sh
edd@rob:~/git/tiledb-soma/apis/r(de/no_r_build_flag)$ ldd /usr/local/lib/R/site-library/tiledbsoma/libs/tiledbsoma.so | grep tiledb
        libtiledbsoma.so => /usr/local/lib/libtiledbsoma.so (0x00007f0be5b1e000)
        libtiledb.so.2.16 => /usr/local/lib/libtiledb.so.2.16 (0x00007f0be4800000)
edd@rob:~/git/tiledb-soma/apis/r(de/no_r_build_flag)$ 
```

There is another minor deficiency in that the library should sport a major and minor number but this is a start.  

@jdblischak : Are that enough initial bread crumps for you to try a Conda build?  (We clearly need work to integrate and polish the suggested changes, I am just content to report that I can make my local system respect a shared library in way that may also be compatible with CI use (at least on Linux, I know too little about macOS).

**Comment by eddelbuettel** [comment]
PS `pkg-config` can now simply ask for both libraries:

```sh
edd@rob:~/git/tiledb-soma/apis/r(de/no_r_build_flag)$ pkg-config --libs tiledbsoma tiledb 
-L/usr/local/lib -ltiledbsoma -ltiledb
edd@rob:~/git/tiledb-soma/apis/r(de/no_r_build_flag)$ 
```

**Comment by eddelbuettel** [comment]
As I suspected above, the C++ library side is not quite there and while it worked for me (on a machine where TileDB Core happens to be a shared library tool) this no longer holds under CI and things fall apart, see https://github.com/single-cell-data/TileDB-SOMA/pull/1280#issuecomment-1520977806  

This has nothing to do with the R package but is infrastructure work required before it can get work -- and that was the the first point I made above.  So at this point this needs someone with cmake chops. I have none.
